add third party code for text-to-sql evaluation

shuaichenchang opened the pull request:
Add third-party code for text-to-sql evaluation. The code is adapted from https://github.com/taoyds/test-suite-sql-eval

neubig commented on explainaboard/third_party/text_to_sql_test_suit_eval/README.md:
```diff
@@ -0,0 +1,144 @@
+# Semantic Evaluation for Text-to-SQL with Test Suites
```
And to clarify, there are some guidelines for adding third-party code here: https://github.com/neulab/ExplainaBoard/tree/main/explainaboard/third_party

shuaichenchang commented:
Modified the Readme and renamed files sql_evaluation.py => evaluation.py, sql_evaluation_test.py => evaluation_test.py to match the code in the third-party repo. We used the different file names for some historical reasons now we changed them back to be the same as the files in the original repo.

shuaichenchang commented:
Change the test criteria in meta_eval_wmt_da_test.py to make it pass the test. Use criteria ci[0]<val<ci[1] instead of hardcode the values of ci[0], ci[1].
e.g.
```
self.assertAlmostEqual(ci[0], 0.6488, 2) 
 self.assertAlmostEqual(ci[1], 0.8999, 2)  
```
to
```
self.assertGreater(val, ci[0]) 
self.assertGreater(ci[1], val) 
```

neubig commented:
Thanks @shuaichenchang ! Could we handle https://github.com/neulab/ExplainaBoard/pull/532/commits/3d5dfeca91c46e6e4faceb0b8a2d18154c7923ad as a separate PR, as it's not related to this addition of third-party code?

pfliu-nlp commented:
> Thanks @shuaichenchang ! Could we handle [3d5dfec](https://github.com/neulab/ExplainaBoard/commit/3d5dfeca91c46e6e4faceb0b8a2d18154c7923ad) as a separate PR, as it's not related to this addition of third-party code?

Maybe approve this case and keep the habit in mind in future PR.

neubig commented:
I'm OK with discussing as part of this thread this time. But I'm actually not sure if the changes to the tests are appropriate, and was hoping to have discussion in the separate thread.

These changes to the unit tests seem to be loosening the check on confidence intervals significantly. I'm wondering if it's appropriate to do so, or if the tests are failing because the code is broken and we need to fix the underlying code. @pfliu-nlp : I think this is code that you implemented, could you confirm that?

pfliu-nlp commented:
>"These changes to the unit tests seem to be loosening the check on confidence intervals significantly. I'm wondering if it's appropriate to do so, or if the tests are failing because the code is broken and we need to fix the underlying code. @pfliu-nlp : I think this is code that you implemented, could you confirm that?"

Yes, @shuaichenchang has consulted me offline; the current modified implementation is something I suggested.

>"not loosen them so much so that they're basically not checking the underlying code."
I think the current implementation is fine. In a lot of other places, we even don't check the ci at all.

neubig commented on integration_tests/meta_eval_wmt_da_test.py:
```diff
@@ -205,8 +205,8 @@ def test_system_level_spearmanr_bootstrap(self) -> None:
         )
-        self.assertAlmostEqual(ci[0], 0.5642, 2)
-        self.assertAlmostEqual(ci[1], 0.9746, 2)
+        self.assertGreater(val, ci[0])
+        self.assertGreater(ci[1], val)
```
```suggestion
        # TODO: This check could be made more rigorous by checking whether the
        #       confidence interval lies within a reasonable range
        self.assertGreater(val, ci[0])
        self.assertGreater(ci[1], val)
```

neubig commented on integration_tests/meta_eval_wmt_da_test.py:
```diff
@@ -188,8 +188,8 @@ def test_sample_level_spearmanr_bootstrap(self) -> None:
         )
-        self.assertAlmostEqual(ci[0], 0.6488, 2)
-        self.assertAlmostEqual(ci[1], 0.8999, 2)
+        self.assertGreater(val, ci[0])
+        self.assertGreater(ci[1], val)
```
```suggestion
        # TODO: This check could be made more rigorous by checking whether the
        #       confidence interval lies within a reasonable range
        self.assertGreater(val, ci[0])
        self.assertGreater(ci[1], val)
```

neubig reviewed (commented):
Thanks for the clarification Pengfei!

> I think the current implementation is fine. In a lot of other places, we even don't check the ci at all.

That's true, but this lack of appropriate checks also [has been a source of bugs in the past](https://github.com/neulab/ExplainaBoard/pull/519).

Are you certain that the implementation is correct, and the failed tests here are just a result of variance in bootstrapping or something similar? If so, then I'd be OK with approving this PR, but suggest that we at least leave a TODO there so we know take a note of this discussion for the future, as I suggested in the revisions here.

neubig commented:
Yeah, I understand the importance of being pragmatic. I created an issue to double-check this though: https://github.com/neulab/ExplainaBoard/issues/537

A difference of 0.08 in the confidence values seems a bit bigger than I would expect from our bootstrapping.

pfliu-nlp commented:
> Yeah, I understand the importance of being pragmatic. I created an issue to double-check this though: #537
> 
> A difference of 0.08 in the confidence values seems a bit bigger than I would expect from our bootstrapping so it'd be nice to know for sure that this is not a bug.

One observation is: it could pass the test in python3.9 while failed in python3.10.

odashi commented:
@neubig @pfliu-nlp Since the issue was created, we need to focus on the original change in this PR.

@shuaichenchang Could you re-request reviews to @neubig and @odashi from the top-right side of this page? it encourages me to continue a review as fast as possible.

shuaichenchang commented:
It seems the ignoring third-party in `.markdownlint.yaml` doesn't skip the tests in the third-party folder. 
https://github.com/neulab/ExplainaBoard/blob/main/.markdownlint.yaml#L27

tetsuok commented:
@shuaichenchang It seems the configuration file needs to be updated. Let me take a look. I'll fix in a separate PR.

tetsuok commented:
@shuaichenchang The fix was merged into `main`. You can merge `main` into this branch to fix markdownlint error.

tetsuok commented:
@shuaichenchang @neubig @pfliu-nlp I created a separate PR #550 to fix `integration_tests/meta_eval_wmt_da_test.py` as soon as possible because I'm observing more frequent test failures in other pull requests, which becomes a drain on our productivity.

pfliu-nlp commented:
I tested this branch locally, it could pass all tests; it's strange that there is a markdown link error here, maybe need to re-commit? (cc @tetsuok )

```
black....................................................................Passed
flake8...................................................................Passed
docstring................................................................Passed
isort....................................................................Passed
Upgrade type hints.......................................................Passed
mypy.....................................................................Passed
markdownlint-cli2........................................................Passed
markdownlint-cli2-fix....................................................Passed
```

tetsuok commented:
@pfliu-nlp It's because this branch is behind `main`. We need merge `main` into this branch and push the commit to the remote.

pfliu-nlp commented:
@tetsuok I see (I assumed that @shuaichenchang has merged the latest main into this PR, but it seemed he didn't), let me do it.
